### PR TITLE
Push screen before OnExiting instead of after

### PR DESCRIPTION
**Pull request by mcendu**
Fixes ppy/osu#8381

Can break some stuff as `next` is no longer the current screen during OnExiting.

### CODE REVIEW

File: osu.Framework/Screens/ScreenStack.cs [hunk 1]
```diff
@@ -257,16 +257,21 @@ private bool exitFrom([CanBeNull] IScreen source, bool shouldFireExitEvent = tru
...
+                // Add the screen back on the stack to allow pushing screens in OnExiting.
+                stack.Push(toExit);
+
                 // if a screen is !ValidForResume, it should not be allowed to block unless it is the current screen (source == null)
                 // OnExiting should still be called regardless.
-                bool blockRequested = toExit.OnExiting(CurrentScreen);
+                bool blockRequested = toExit.OnExiting(next);
 
                 if ((source == null || toExit.ValidForResume) && blockRequested)
                 {
```
Comment (iiSaLMaN) [inline osu.Framework/Screens/ScreenStack.cs]:
```suggestion
```

File: osu.Framework/Screens/ScreenStack.cs [hunk 2]
```diff
@@ -257,16 +257,21 @@ private bool exitFrom([CanBeNull] IScreen source, bool shouldFireExitEvent = tru
...
-                    // If the exit event gets cancelled, add the screen back on the stack.
-                    stack.Push(toExit);
                     return true;
                 }
```
Comment (iiSaLMaN) [inline osu.Framework/Screens/ScreenStack.cs]:
```suggestion
```

**Comment by peppy** [comment]
tests would be great!

**Comment by mcendu** [comment]
I am not sure about how to refer to the newly created object inside the object initializer itself, without inheritance or triggering CI error.

**Comment by smoogipoo** [comment]
Can't push to your branch but you can rewrite it as:
```csharp
pushAndEnsureCurrent(() =>
{
    screen1 = new TestScreen(id: 1);
    screen1.Exiting = () => 
    {
        screen1.Push(new TestScreen(id: 2));
        return true;
    };

    return screen1;
}
```

**Pull request merged**